core/src/main/scala/org/bykn/bosatsu/Evaluation.scala

```diff
@@ -508,28 +508,12 @@ case class Evaluation[T](pm: PackageMap.Typed[T], externals: Externals) {
          val pack = pm.toMap.get(p).getOrElse(sys.error(s"cannot find $p, shouldn't happen due to typechecking"))
          Scoped.const(getValue(pack, ident))
        case App(fn, arg, _, _) =>
-         fn match {
-           case AnnotatedLambda(name, _, fn, _) =>
-             // f(\x -> res)(y) = let x = y in res
-             val argE = recurse(arg)
-             val fnE = recurse(fn)
-
-             argE.letNameIn(name, fnE)
-           case fn =>
-             val efn = recurse(fn)
-             val earg = recurse(arg)
-
-             efn.applyArg(earg)
-         }
+         val efn = recurse(fn)
+         val earg = recurse(arg)
+
+         efn.applyArg(earg)
        case AnnotatedLambda(name, _, expr, _) =>
-         expr match {
-           case App(f, Var(None, name2, _, _), _, _) if name2 == name =>
```

Comment:
this rule is too simplistic... the TypedExpr.normalization code has a better implementation that (hopefully) is bug-free.

The problem is `\x -> f(x)(x)` you can't replace this with `f(x)` since now `x` is unbound.